File: liqp-core/src/main/java/liqp/tags/Assign.java
package liqp.tags;

import java.util.List;
import liqp.tag.LTag;
import liqp.filter.FilterChain;
import liqp.filter.FilterParams;
import liqp.filter.FilterInstance;
import liqp.nodes.FilterNode;
import liqp.node.LNode;
import liqp.context.LContext;
import one.util.streamex.StreamEx;

public class Assign extends LTag {

  /*
   * Assigns some value to a variable
   */
  @Override
  public Object render(LContext context, LNode... nodes) {
    final List<FilterInstance> filters = StreamEx.of(nodes)
          .filter(node -> node instanceof FilterNode)
          .map(node -> (FilterNode) node)
          .map(node -> new FilterInstance(node.getFilter(), new FilterParams(node.getParams())))
          .toImmutableList();
    final FilterChain chain = new FilterChain(context, filters, ref -> {
      LNode expression = nodes[1];
      ref.set(expression.render(context));
      return null;
    });

    final Object finalValue = chain.processFilters();
    String id = String.valueOf(nodes[0].render(context));
    // Assign causes variable to be saved "globally"
    context.setRoot(id, finalValue);
    return "";
  }
}

File: liqp-core/src/main/kotlin/liqp/tags/Assign.kt
package liqp.tags

import liqp.ControlResult
import liqp.context.LContext
import liqp.filter.FilterChain
import liqp.filter.FilterInstance
import liqp.filter.FilterParams
import liqp.node.LNode
import liqp.nodes.FilterNode
import liqp.tag.LTag

class Assign : LTag() {

  /**
   * Assigns some value to a variable
   */
  override fun render(context: LContext, vararg nodes: LNode): Any? {
    val filters = nodes
        .filter { node -> node is FilterNode }
        .map { node -> node as FilterNode }
        .map { node -> FilterInstance(node.filter, FilterParams(context, node.params)) }
    val chain = FilterChain(context, filters) {
      val expression = nodes[1]
      result = expression.render(context)
      ControlResult.NO_CONTENT
    }

    val finalValue = chain.processFilters()
    val id = nodes[0].render(context).toString()
    // Assign causes variable to be saved "globally"
    context.setRoot(id, finalValue!!)
    return null
  }
}
